kratos/utilities/parallel_utilities.h

```diff
@@ -0,0 +1,364 @@
+
+    virtual ~BlockPartition() {}
+
+    //version enabling reduction
+    //simple version
+    template <class TUnaryFunction>
+    inline void for_each(TUnaryFunction&& f)
+    {
+        #pragma omp parallel for
+        for(int i=0; i<mNchunks; ++i)
+        {
+            for (int k = mBlockPartition[i]; k < mBlockPartition[i+1]; ++k)
+            {
+                f(*(mit_begin + k)); //note that we pass a reference to the value, not the iterator
+            }
+        }
+    }
+
+    //version with reduction
+    template <class TReducer, class TUnaryFunction>
+    inline TReducer for_reduce(TUnaryFunction &&f)
+    {
+        TReducer global_reducer;
+        #pragma omp parallel for
+        for(int i=0; i<mNchunks; ++i)
+        {
+            TReducer local_reducer;
+            for (auto k = mBlockPartition[i]; k < mBlockPartition[i+1]; ++k)
+            {
+                local_reducer.LocalMerge(f(*(mit_begin + k)));
+            }
+            global_reducer.ThreadSafeMerge(local_reducer);
+        }
+        return global_reducer;
+    }
+
+private:
+
+    TIteratorType mit_begin, mit_end;
+    int mNchunks;
+    std::array<int, max_allowed_threads> mBlockPartition;
+
+};
+
+//***********************************************************************************
+//***********************************************************************************
+//***********************************************************************************
+///This class is useful for iteration over containers
+template<class TIndexType>
+class IndexPartition
+{
+public:
+
+    static constexpr int max_allowed_threads = 128;
+
+    IndexPartition(TIndexType Size,
+                   int Nchunks = omp_get_max_threads())
+        : mSize(Size), mNchunks(Nchunks)
+    {
+        int mBlockPartitionize = mSize / mNchunks;
+        mBlockPartition[0] = 0;
+        mBlockPartition[mNchunks] = mSize;
+        for(int i = 1; i < mNchunks; i++)
+            mBlockPartition[i] = mBlockPartition[i-1] + mBlockPartitionize ;
+
+    }
+
+    virtual ~IndexPartition() {}
+
+    //pure c++11 version (can handle exceptions)
+    template <class TUnaryFunction>
+    inline void for_pure_c11(TUnaryFunction &&f)
+    {
+        KRATOS_TRY
+        std::vector< std::future<void> > runners(mNchunks);
+        const auto& partition = mBlockPartition;
+        for(int i=0; i<mNchunks; ++i)
+        {
+            runners[i] = std::async(std::launch::async, [&partition, i,  &f]()
+                {
+                    for (auto k = partition[i]; k < partition[i+1]; ++k)
+                        f(k);
+                });
+        }
+
+        //here we impose a syncronization and we check the exceptions
+        for(int i=0; i<mNchunks; ++i)
+        {
+            try {
+                    runners[i].get();
+            } catch(const std::exception& e) {
+                std::cout << std::endl << "THREAD number: " << i << " caught exception " << e.what() << std::endl;
```

Comment:
Should not you rethrow the exception here? Just ignoring it does not seem like a nice thing to do.

Reply:
mmm. I am out of my depths here.

so you would have:

~~~cpp
{
     KRATOS_ERROR << std::endl << "THREAD number: " << i << " caught exception " << e.what() << std::endl;
~~~

Reply:
If you want to output the error here, then just add
```cpp
throw;
```
below. If the output is not required, then just get rid of the `try .. catch` block, `runners[i].get()` will through for you.

Reply:
@pooyan-dadvand any hint on this?

Reply:
ok, i have committed this, unfortunately this also means that i will have to comment out the test to be able to pass

Reply:
Usually, test frameworks allow testing for expected exceptions. For example, https://www.boost.org/doc/libs/1_31_0/libs/test/doc/components/test_tools/reference/BOOST_CHECK_THROW.html

Reply:
yes, also the Kratos framework allows to do that.
the problem here is that the error message contains the thread number which will depend on the system the test is being run on

Reply:
I think you can create the test such that always thread 0 throws and check this in the error message
Do you even have to give the entire string of the error message? Cannot it only contain a substring?

Reply:
ok, it is a good suggestion, i will make the test so that the error is thrown always by the first thread and capture it